### DISCUSSION TITLE
Reverse motor in code

### DISCUSSION THREAD

**Discussion by dwfunk4475** [gizmo-platform/gizmo#36] [Q&A] [answered]
How do you reverse a motor in code?

Is there a list of gizmo commands available somewhere we've over looked?

Thanks!

-david

**Comment by the-maldridge** [comment, marked as answer]
Reversing a motor isn't gizmo specific, so there wasn't anything in our docs for you to overlook.  Based on your other thread I'll assume this is in the context of Arduino.   In the Arduino environment using VeX MC-29 motor controllers motors are modeled as servos in continuous rotation mode.  The main arduino docs for the servo libarary can be found [here](https://docs.arduino.cc/libraries/servo/), but the gist of the solution is that the servo `write` function takes a value between 0 and 180, where 90 is stopped, 0 is full speed in one direction, and 180 is full speed in the other.

If you just want to reverse which direction is considered to be "forward" I would suggest you might want to do that electrically by just flipping the polarity on the motor in question.

**Reply by dwfunk4475** [reply]
Thanks! Didn't pickup on treating this controller like a servo.
In other applications, there was usually some configuration using, forward/reverse, or true/false to set a motor forward or reverse. Was trying to figure out where that might be in this context.

**Reply by the-maldridge** [reply]
NP.  The VeX MC-29 motor controllers are somewhat unique in how exactly they behave like servos from a controls standpoint.  You're absolutely right that other motor controllers usually will need a direction bit.  In the Arduino world, this is usually handled for you by centering the motor's total throttle range on zero and using positive values for one direction and negative values for the other.

**Comment by plurt97** [comment]
I believe the Arduino sample code from BEST has a bug in it where they send the same command to the left and right motor causing your robot to go around in circles.  The solution is to either reverse the wires on the motor (a really bad idea, don't do it), or remap the right motor.  While I don't understand the concept of the 0 to 180 with 90 in the middle (all other robot code I've programmed in use -127 to 127 with 0 meaning stop and I received furled brows from my students as well trying to teach them this.:), to reverse it (ala the "reverse" command or "false"), you need to remap the values using the map() function.    speed = map(speed, 0, 180, 180, 0);
Hope this answers your question (if not, I apologize for the wasted time you spent reading it. :-)

**Reply by the-maldridge** [reply]
Its not a "really bad idea" to reverse the wires.  DC motors are not polarity sensitive, and unless you are extremely diligent about defining which direction it should spin when a positive voltage is applied to a specified terminal, the concept of "forward" is undefined.

`map()` is overkill when you want to reverse the range of a fixed positive integer space, just subtract your value from the max of the range, so `speed = 180 - speed` will work just fine.  As to why its that range at all, this is an arbitrary decision made by the authors of the Servo library for Arduino, and has to do with historical reasons beyond the scope of this thread.  Most serious motor controllers stopped using pure PWM as an input control scheme quite some time ago, so this is all moot anyway since more modern DC motor controllers use a well defined serial or I2C protocol which do not rely on sentinel values or in-band signaling to convey meaning to things like direction.

**Reply by dwfunk4475** [reply]
Actually, the sample code does not have a bug in it. The sample code is correctly telling both motors to go forwards.

All of my students miss this next point, as do most of my mentors.
Look carefully at your motors. For the sake of this explaination, 'forward' equals clockwise rotation of the motor. Hence right side of robot moves forwards.

Now here's the simple little thing that always gets overlooked. Look at the left motor, which is mounted 180 degrees from the right motor. Catch it yet?😃 it's easy to miss. That motor is also turning 'forward' which is a clockwise rotation, and that makes the left side of the robot move backwards!

The solution is as previously mentioned, reverse the DC wires to the motor. There is no positive or negative on DC motors, the red/black wires are for convenience. Switch them at the white terminal block.

Never, ever reverse the 3pin connector of the motor controller on gizmo. It's catastrophic !
